src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -244,13 +567,132 @@ impl Crumbable for crate::Block<Ast> {
     }
 }
 
+impl Crumbable for crate::TextLineFmt<Ast> {
+    type Crumb = TextLineFmtCrumb;
+
+    fn get(&self, crumb:&Self::Crumb) -> FallibleResult<&Ast> {
+        let segment = self.text.get(crumb.segment_index);
+        let segment = segment.ok_or(IndexOutOfBounds("text segment".into()))?;
+        if let crate::SegmentFmt::SegmentExpr(expr) = segment {
+            expr.value.as_ref().map(|ast| ast).ok_or_else(|| NotPresent("expression".into()).into())
+        } else {
+            Err(NotPresent("segment expression".into()).into())
+        }
+    }
+
+    fn set(&self, crumb:&Self::Crumb, new_ast:Ast) -> FallibleResult<Self> {
+        let mut text = self.clone();
+        let segment = text.text.get_mut(crumb.segment_index);
+        let segment = segment.ok_or(IndexOutOfBounds("text segment".into()))?;
+        if let crate::SegmentFmt::SegmentExpr(expr) = segment {
+            expr.value = Some(new_ast);
+            Ok(text)
+        } else {
+            Err(NotPresent("segment expression".into()).into())
```

Comment:
Is the suggestion to replace segment's `index - 1` if `index` is not present? I don't quite understand how it could be useful.

Reply:
No, I meant "previous" as in "before being replaced" — that is the case when segment exists but is not an expression segment.
THen by setting AST on this it could just replace the non-expression segment with expression segment with provided AST.

But let's leave this as-is for now, I'm mentioning this more to be aware in a future that if use-case appears, we know that we can adjust crumbs here.